core/src/main/scala/cilib/Fitness.scala

```diff
@@ -2,82 +2,75 @@ package cilib
 
 import scalaz._
 import scalaz.Ordering._
-import scalaz.std.anyVal._
-import scalaz.syntax.equal._
+import Scalaz._
 
-sealed trait Fit {
+sealed abstract class Fit {
   def fold[Z](penalty: Adjusted => Z, valid: Feasible => Z, infeasible: Infeasible => Z): Z =
     this match {
-      case p @ Adjusted(_, _)   => penalty(p)
-      case v @ Feasible(_)      => valid(v)
-      case x @ Infeasible(_, _) => infeasible(x)
+      case p @ Adjusted(_, _) => penalty(p)
+      case v @ Feasible(_)    => valid(v)
+      case x @ Infeasible(_)  => infeasible(x)
+    }
+
+  def adjust(f: Double => Double): Fit =
+    this match {
+      case Adjusted(_, _)    => this
+      case x @ Feasible(v)   => Adjusted(\/.left(x), f(v))
+      case x @ Infeasible(v) => Adjusted(\/.right(x), f(v))
     }
 }
 
 final case class Feasible(v: Double) extends Fit
-final case class Infeasible(v: Double, violations: Int) extends Fit {
-  def adjust(f: Double => Double) =
-    Adjusted(this, f(v))
-}
-final case class Adjusted private[cilib] (original: Infeasible, adjust: Double) extends Fit
-@annotation.implicitNotFound("No instance of Fitness[${F},${A}] is available in current scope.")
-trait Fitness[F[_], A] {
-  def fitness(a: F[A]): Option[Objective[A]]
+final case class Infeasible(v: Double /*, violations: Int*/ ) extends Fit
+final case class Adjusted private[cilib] (original: Feasible \/ Infeasible, adjust: Double)
+    extends Fit
+@annotation.implicitNotFound(
+  "No instance of Fitness[${F},${A},${B}] is available in current scope.")
+trait Fitness[F[_], A, B] {
```

Comment:
When will B be different to A?

Reply:
It shouldn't be, but if you look at the typeclass, the output is defined to be `Option[Objective[B]]`, so the additional type parameter allows a different type for the output, vs the input. We mostly use it in the tests, but in terms of the real usages, `A =:= B` in almost all cases.

Reply:
But in Objective, the parameter is used to define the Constraint, and in Constraint the parameter is used to define the input again, or is that not the case?

Reply:
I don't think that this is the case, no.